File: core-vm/src/native/wonka/vm/Finalizer.c
/**************************************************************************
* Copyright (c) 2001, 2002, 2003 by Acunia N.V. All rights reserved.      *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Philips site 5, box 3       info@acunia.com                           *
*   3001 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
**************************************************************************/

/*
** $Id: Finalizer.c,v 1.2 2004/07/18 18:43:49 cvs Exp $
*/

#include "clazz.h"
#include "core-classes.h"
#include "fifo.h"
#include "heap.h"
#include "oswald.h"

w_instance Finalizer_nextFinalizee(JNIEnv *env, w_instance theFinalizer) {
  w_instance nextFinalizee;

  if (!finalizer_fifo) {

    return NULL;

  }

  x_mutex_lock(finalizer_fifo_mutex, x_eternal);
  nextFinalizee = getFifo(finalizer_fifo);
  if (nextFinalizee) {
    woempa(7, "Took %j from finalizer_fifo, now has %d elements\n", nextFinalizee, finalizer_fifo->numElements);
  }
  x_mutex_unlock(finalizer_fifo_mutex);

  return nextFinalizee;
}

w_instance Finalizer_nextEnqueueee(JNIEnv *env, w_instance theFinalizer) {
  w_instance nextEnqueueee;

  if (!enqueue_fifo) {

    return NULL;

  }

  x_mutex_lock(enqueue_fifo_mutex, x_eternal);
  nextEnqueueee = getFifo(enqueue_fifo);
  if (nextEnqueueee) {
    woempa(7, "Took %j from enqueue_fifo, now has %d elements\n", nextEnqueueee, enqueue_fifo->numElements);
  }
  x_mutex_unlock(enqueue_fifo_mutex);

  return nextEnqueueee;
}

/**
 ** When the object has been finalized we clear the FINALIZING flag.
 */
void Finalizer_finalized(JNIEnv *env, w_instance theFinalizer, w_instance finalizee) {
  unsetFlag(instance2object(finalizee)->flags, (O_FINALIZING));
}

/**
 ** When the object has been enqueued we clear the ENQUEUEING flag.
 */
void Finalizer_enqueued(JNIEnv *env, w_instance theFinalizer, w_instance enqueuee) {
  unsetFlag(instance2object(enqueuee)->flags, (O_ENQUEUEING));
}

File: Mika/trunk/core-vm/java/wonka/vm/Finalizer.java
/**************************************************************************
* Copyright  (c) 2001 by Acunia N.V. All rights reserved.                 *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Vanden Tymplestraat 35      info@acunia.com                           *
*   3000 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
*                                                                         *
* Modifications copyright (c) 2004 by Chris Gray, /k/ Embedded Java       *
* Solutions. All rights reserved.                                         *
*                                                                         *
**************************************************************************/

/*
** $Id: Finalizer.java,v 1.5 2005/09/03 12:11:33 cvs Exp $
*/

package wonka.vm;

public final class Finalizer implements Runnable {

  private static Finalizer theFinalizer;

  private static final int IDLE_WAIT_MILLIS = 1000;

  private Object[] empty;
  private java.lang.reflect.Method finalizer_method;
  private java.lang.reflect.Method enqueue_method;

  /**
   ** Our very private constructor.  We look up the methodID's for the
   ** methods Reference/enqueue() and Object/finalize() and start our
   ** thread running.
   */
  private Finalizer() {

    try {
      Class java_lang_Object = Class.forName("java.lang.Object");
      finalizer_method = java_lang_Object.getDeclaredMethod("finalize", new Class[0]);
      finalizer_method.setAccessible(true);

      Class java_lang_ref_Reference = Class.forName("java.lang.ref.Reference");
      enqueue_method = java_lang_ref_Reference.getDeclaredMethod("enqueue", new Class[0]);
      enqueue_method.setAccessible(true);
    }
    catch (Exception e) {
      e.printStackTrace();
    }
    empty = new Object[0];
    Thread thread = new Thread(this, "Confessor");
    thread.setPriority(10);
    thread.setDaemon(true);
    thread.start();

  }

  /**
   ** Return the solitary instance of Finalizer.
   */
  public static synchronized Finalizer getInstance() {
    if (theFinalizer == null) {
      theFinalizer = new Finalizer();
    }

    return theFinalizer;
  }

  /**
   ** The run() method polls the two internal fifo's and calls enqueue()
   ** or finalize() as appropriate.
   */
  public void run() {
    boolean idle;
    boolean verbose;
    int finalized_count = 0;
    int enqueued_count = 0;
    Object o;

    // Give GC thread time to start up, otherwise nextFinalizee() crashes (kludge).
    try {
      synchronized(this) {
        wait(IDLE_WAIT_MILLIS);
      }
    }
    catch (InterruptedException ie) {
    }

    verbose = (System.getProperty("mika.verbose", "").indexOf("gc") >= 0);

    while(true) {
      idle = true;
      synchronized(this) {
        o = nextFinalizee();
      }

      if (o == null) {
        synchronized(this) {
          notifyAll();
        }
      }
      else {
        idle = false;
        if (verbose) {
          System.err.println("GC: finalizing " + o);
        }
        try {
          finalizer_method.invoke(o, empty);
        }
        catch (Throwable t) {
          t.printStackTrace();
        }
        finalized(o);
        ++finalized_count;
      }

      synchronized(this) {
        o = nextEnqueueee();
      }

      if (o == null) {
        synchronized(this) {
          notifyAll();
        }
      }
      else {
        idle = false;
        if (verbose) {
          System.err.println("GC: enqueuing " + o);
        }
        try {
          enqueue_method.invoke(o, empty);
        }
        catch (Throwable t) {
          t.printStackTrace();
        }
        enqueued(o);
        ++enqueued_count;
      }

      if (idle) {
        finalized_count = 0;
        enqueued_count = 0;
        synchronized(this) {
          try {
            wait(IDLE_WAIT_MILLIS);
          }
          catch (Throwable t) {
            t.printStackTrace();
          }
        }
      }
    }
  }

  /**
   ** Poll the enqueue() fifo.
   */
  private native Object nextEnqueueee();

  /**
   ** Poll the finalize() fifo.
   */
  private native Object nextFinalizee();

  /**
   ** Tell GC that object has been finalized.
   */
  private native void finalized(Object o);

  /**
   ** Tell GC that object has been enqueued.
   */
  private native void enqueued(Object o);
}

